### PR TITLE
backend login/register/check

### PR DESCRIPTION
**Pull request by ubungmeister**

### CODE REVIEW

File: backend/src/helpers/functions.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
+import {NextFunction, Response, Request} from "express";
```
Comment (penkobor) [inline backend/src/helpers/functions.ts]: I think this file should be separated into multiple different files (now probably 2) and the folder they are in should be called something like `ViewModels`. 

As far as I know there are many different design patterns for BE application, but the most famous are MVC and MVVM and they all follow the same idea: 
- to have your interface (your api or you routes) in one place. Its basically the place where the interface for accessing your BE from outer world is defined. You have it all in `backend/index.ts`
- to have `Models` its where Entities are defined and probably linked to DB tables
- to have a place where your "business logic" is happening. In your case its this `functions.ts` file. 

You basically have it devided in a similar fashion, but its all a bit mixed up. 

Things I would suggest: 
- rename `helpers` folder into something that makes more sense. I would suggest `viewmodels` (MVVM). But maybe its not the best option, try to google MVVM and MVC patterns. Basically that is the folder where your business logic will be.
- inside this folder separate this `functions.ts` into `PostVM.ts` and `UserVM.ts` and move corresponding logic to this files. Idea is to keep Models and "business logic" logically grouped together. So you don't have one file responsible for everything, but you have 1 file with all logic for 1 model

File: backend/src/helpers/functions.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        res.json({...userData, token})
+    } catch (err) {
+        console.log(err)
+        res.status(500).json({
+            message: 'Not able to sing in'
+        })
+    }
+}
+
+export const checkAuthorisation = async (req: Request, res: Response) => {
```
Comment (penkobor) [inline backend/src/helpers/functions.ts]: seems redundant, Why do you need to check if the person is present in DB ?

Comment (penkobor) [reply backend/src/helpers/functions.ts]: more about this in comments on `index.ts`

File: backend/index.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    body('email').isEmail(),
+    body('password', "mi222n 5").isLength({min: 6}),
+    body('fullName').isLength({min: 3}),
+    register
+);
+app.post('/auth/login',
+    body('email').isEmail(),
+    body('password', "mi222n 5").isLength({min: 6}),
+    login)
+app.get('/auth/me',checkAuth,checkAuthorisation)
```
Comment (penkobor) [inline backend/index.ts]: so I think this route is completely redundant. 
The standard flow looks like this:
1. User registers - he gets appended to DB, no need to give user token at this point
2. User logs in - then he is issued a token
3. On **each further request** (this is the important part) you need to check if token is valid. That is basically the point of this token. And if the token is not valid BE should refuse to answer and return an error that token is invalid. 

So there is no reason to have an endpoint just for checking the token. 
If you called it `userInfo` that would make more sense, then you could tell its an endpoint for getting user information for maybe displaying it somewhere in the application

File: backend/index.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+//upload photo
+app.post('/upload' ,upload.single('image'),(req:Request,res:Response)=>{
+    res.json({
+        url:`/upload/${req.file?.originalname}`
+    })
+})
+//post
+app.post('/posts',
```
Comment (penkobor) [inline backend/index.ts]: also I believe you could follow the same idea of MVC or MVVM and create some "Controllers". 
As I see it in "src" you would have another folder with "controllers" where endpoints for each Model is defined separately. Not sure how to make it properly in Node application, but the easiest option that comes to find is to create `src/controllers/PostsController.ts` and there defined a function `createPostEndpoints(app)` which will have logic connected to posts, and do the same for other entities. 

If you do it like this then your index.ts would look like. 

```typescript
...
const app = express()
app.use(express.json())
app.use('/upload', express.static('upload')) //catch GET request to static files
...
createAuthEndpoints(app);
createPostsEndpoints(app);
createUploadEndpoints(app);
```

Naming could be different, but I hope that the point that I'm trying to make is understandable.

File: backend/index.ts (hunk 3 above)
Comment (ubungmeister) [reply backend/index.ts]: fixed, I keep it for displaying user information in the frontend part. Change the name to userInfo so it makes more sense

File: backend/index.ts (hunk 4 above)
Comment (ubungmeister) [reply backend/index.ts]: fixed

File: backend/src/helpers/functions.ts (hunk 1 above)
Comment (ubungmeister) [reply backend/src/helpers/functions.ts]: helpers function separate to controllers,
keep models, 
separate routes.

**Pull request merged**